### PR TITLE
Properly handle node inputs for materials

### PR DESCRIPTION
**Pull request by JGamache-autodesk**

### CODE REVIEW

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 1]
```diff
@@ -776,6 +776,18 @@ namespace
...
+                    {
+                        auto outputName = std::string("OUT_") + input->getNodeName() +
+                                          "_" + input->getOutputString();
+                        auto output = doc->addOutput(outputName, input->getType());
+                        output->setNodeName(input->getNodeName());
+                        output->setAttribute("output", input->getOutputString());
+                        bindInput->setOutputString(outputName);
+                    }
+                }
+                else if (input->hasNodeGraphName())
```
Comment (JGamache-autodesk) [inline source/MaterialXRuntime/RtFileIo.cpp]: Existing code works for nodegraph elements and was kept.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 2]
```diff
@@ -776,6 +776,18 @@ namespace
             {
                 BindInputPtr bindInput = shaderRef->addBindInput(input->getName(), input->getType());
                 if (input->hasNodeName())
+                {
```
Comment (JGamache-autodesk) [inline source/MaterialXRuntime/RtFileIo.cpp]: Connections at the node element level require an intermediate output element,

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 3]
```diff
@@ -1002,7 +1015,7 @@ namespace
                 if (output && output.getType() == MATERIAL_TYPE_STRING && writeOptions &&
                     writeOptions->materialWriteOp & RtWriteOptions::MaterialWriteOp::WRITE_MATERIALS_AS_ELEMENTS)
                 {
-                    writeMaterialElement(mxNode, doc, writeOptions);
+                    materialElements.push_back(mxNode);
```
Comment (JGamache-autodesk) [inline source/MaterialXRuntime/RtFileIo.cpp]: Delay element conversion until all XML elements are persisted.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 4]
```diff
@@ -776,6 +776,18 @@ namespace
             {
                 BindInputPtr bindInput = shaderRef->addBindInput(input->getName(), input->getType());
                 if (input->hasNodeName())
+                {
+                    if (input->hasOutputString())
+                    {
+                        auto outputName = std::string("OUT_") + input->getNodeName() +
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: minor: can this be const?

**Comment by feldstj** [comment]
I'm still not sure I see the purpose of top-level outputs, I hope we can do away with them at some point. But for now, since they do exist, this makes sense. Thanks for adding this, just a couple of minor comments.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 5]
```diff
@@ -776,6 +776,18 @@ namespace
...
+                                          "_" + input->getOutputString();
+                        auto output = doc->addOutput(outputName, input->getType());
+                        output->setNodeName(input->getNodeName());
+                        output->setAttribute("output", input->getOutputString());
```
[the remark this answers is not in the document]
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp]: I'm not sure what the situation is but this is creating an invalid bindinput. (You can't have syntax "nodename=").  
The spec says that "nodegraph=" refers to a nodegraph element with "output=" being optional

> - nodegraph (string, optional): the name of the nodegraph element whose output will be bound
> - output (string, optional): the name of the nodegraph output that will be bound to this input
 

Yes, the nomenclature is confusing. If it's a free floating node then it's not acceptable (for now) as being saveable to a document and still be "valid".

Maybe a test case would help to clarify here. Thanks.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 6]
```diff
@@ -775,13 +776,25 @@ namespace
             for (InputPtr input : surfaceShader->getActiveInputs())
             {
                 BindInputPtr bindInput = shaderRef->addBindInput(input->getName(), input->getType());
-                if (input->hasNodeName())
+                if (input->hasNodeName() && input->hasOutputString())
                 {
-                    if (input->hasOutputString())
+                    if (doc->getNodeGraph(input->getNodeName()))
```
Comment (bernardkwok) [inline source/MaterialXRuntime/RtFileIo.cpp]: Thanks for managing the lack of "nodename" support (one thing upgrading to a node tries to address).

**Pull request merged**